lib/private/Files/External/InvalidStorage.php

```
@@ -0,0 +1,31 @@
+<?php
+/**
+ * @author Vincent Petry <pvince81@owncloud.com>
+ *
+ * @copyright Copyright (c) 2016, ownCloud GmbH.
+ * @license AGPL-3.0
+ *
+ * This code is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License, version 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License, version 3,
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>
+ *
+ */
+
+namespace OC\Files\External;
+
+use OCP\Files\StorageNotAvailableException;
+use OC\Files\Storage\FailedStorage;
+
+class InvalidStorage extends FailedStorage {
+	public function __construct($params) {
+		throw new StorageNotAvailableException();
```

This seems too "artificial" for me. Creating a class that would throw an exception in the constructor implies that you want to throw the exception in the first place, which makes me wonder why do we need to create a class for that.

It probably makes more sense if the exception is thrown in specific methods even if those methods are almost all of them. My point is that the InvalidStorage should behave as another storage, but that exception doesn't allow it.

---

The actual reason is that at the location where we specify the class name we don't actually instantiate it. Remember that the external storage backend class must specify a name for the storage class to be instantiated, it doesn't instantiate it itself, but MountPoint->createStorage will instantiate it later. Throwing an exception in the constructor is to make sure this never happens.

---

My worry is that this is something that wasn't originally planned.

Due to there isn't an oficial storage factory, other apps like WND have to create the storages on their own. Throwing the exception like that will likely affect them even if the storage won't be used. I'll have to recheck the behaviour.
Creating the storage and throwing the exception in the rest of the methods seems a more compatible option.

---

ok, it seems WND will check the backend class and won't try to create an storage instance if it isn't for the WND backend

---

There is a `StorageFactory` class: https://github.com/owncloud/core/blob/v10.0.3/lib/private/Files/Storage/StorageFactory.php and here is the code that instantiates storages: https://github.com/owncloud/core/blob/v10.0.3/lib/private/Files/Mount/MountPoint.php#L138 (the $loader is actually an instance of `StorageFactory`.

In general third party apps should never instantiate the storage class themselves.

The use case of this InvalidStorage class should also never be accessible for apps as it's designed for use by core when accessing a storage for which the actual storage class does not exist due to the app being disabled.

---

I'm out of arguments against it, but it still feels like adding tomato to the spaghetti code... 🤕

---

Hahaha, it does.

Unfortunately the backends came after the storages. If we designed storages to have backends first, we could have designed it so that the backend is responsible of instantiating storages, in which case this mess above wouldn't required.